File: implementations/nodejs/velora.ts
/**
 * Velora Core - TypeScript/Node.js Implementation
 * AI Interoperability Layer for seamless integration
 */

import { EventEmitter } from 'events';
import * as crypto from 'crypto';
import * as net from 'net';
import { v4 as uuidv4 } from 'uuid';
import * as xml2js from 'xml2js';
import * as iconv from 'iconv-lite';

/**
 * Main Velora Core class
 */
export class VeloraCore extends EventEmitter {
    private instanceId: string;
    private startTime: Date;
    private isRunning: boolean;
    
    // Components
    private protocolManager?: ProtocolManager;
    private agentManager?: AgentManager;
    private securityManager?: SecurityManager;
    private legacyBridge?: LegacyBridge;
    private dataManager?: DataManager;
    
    // Configuration
    private config: VeloraConfig;
    
    constructor(config?: VeloraConfig) {
        super();
        this.config = config || VeloraConfig.getDefault();
        this.instanceId = uuidv4();
        this.startTime = new Date();
        this.isRunning = false;
        
        console.log(`Velora Core initialized: ${this.instanceId}`);
    }
    
    /**
     * Initialize all components
     */
    async setup(): Promise<void> {
        try {
            console.log('Setting up Velora components...');
            
            // Initialize Security Manager first
            this.securityManager = new SecurityManager(this.config);
            await this.securityManager.initialize();
            
            // Initialize Protocol Manager
            this.protocolManager = new ProtocolManager(this.config);
            await this.protocolManager.initialize();
            
            // Initialize Agent Manager
            this.agentManager = new AgentManager(this.config, this.protocolManager);
            await this.agentManager.initialize();
            
            // Initialize Legacy Bridge
            this.legacyBridge = new LegacyBridge(this.config);
            await this.legacyBridge.initialize();
            
            // Initialize Data Manager
            this.dataManager = new DataManager(this.config);
            await this.dataManager.initialize();
            
            console.log('Velora setup completed');
            
        } catch (error) {
            console.error('Failed to setup Velora:', error);
            throw new VeloraException('Setup failed', error);
        }
    }
    
    /**
     * Start all components
     */
    async start(): Promise<void> {
        if (this.isRunning) return;
        
        console.log('Starting Velora...');
        
        // Start all components
        await this.protocolManager?.start();
        await this.agentManager?.start();
        await this.securityManager?.start();
        await this.legacyBridge?.start();
        await this.dataManager?.start();
        
        this.isRunning = true;
        this.emit('started');
        console.log('Velora started successfully');
    }
    
    /**
     * Stop all components
     */
    async stop(): Promise<void> {
        if (!this.isRunning) return;
        
        console.log('Stopping Velora...');
        
        // Stop all components in reverse order
        await this.dataManager?.stop();
        await this.legacyBridge?.stop();
        await this.agentManager?.stop();
        await this.protocolManager?.stop();
        await this.securityManager?.stop();
        
        this.isRunning = false;
        this.emit('stopped');
        console.log('Velora stopped');
    }
    
    /**
     * Perform health check
     */
    async healthCheck(): Promise<HealthCheckResult> {
        const uptime = (Date.now() - this.startTime.getTime()) / 1000;
        
        return {
            status: this.isRunning ? 'healthy' : 'stopped',
            instanceId: this.instanceId,
            uptime,
            components: {
                protocols: this.protocolManager ? 'active' : 'inactive',
                agents: this.agentManager ? 'active' : 'inactive',
                security: this.securityManager ? 'active' : 'inactive',
                legacy: this.legacyBridge ? 'active' : 'inactive',
                data: this.dataManager ? 'active' : 'inactive'
            }
        };
    }
    
    /**
     * Get system status
     */
    getStatus(): SystemStatus {
        return {
            instanceId: this.instanceId,
            isRunning: this.isRunning,
            startTime: this.startTime,
            uptime: (Date.now() - this.startTime.getTime()) / 1000,
            environment: this.config.environment,
            version: this.config.version
        };
    }
    
    // Getters
    get protocols(): ProtocolManager | undefined { return this.protocolManager; }
    get agents(): AgentManager | undefined { return this.agentManager; }
    get security(): SecurityManager | undefined { return this.securityManager; }
    get legacy(): LegacyBridge | undefined { return this.legacyBridge; }
    get data(): DataManager | undefined { return this.dataManager; }
}

/**
 * Protocol Manager - Handles UAICP and ANP
 */
export class ProtocolManager {
    private config: VeloraConfig;
    private uaicp?: UAICP;
    private anp?: ANP;
    
    constructor(config: VeloraConfig) {
        this.config = config;
    }
    
    async initialize(): Promise<void> {
        this.uaicp = new UAICP(this.config.agentId);
        this.anp = new ANP(this.config.agentId);
    }
    
    async start(): Promise<void> {
        await this.uaicp?.start();
        await this.anp?.start();
    }
    
    async stop(): Promise<void> {
        await this.anp?.stop();
        await this.uaicp?.stop();
    }
    
    getUAICP(): UAICP | undefined { return this.uaicp; }
    getANP(): ANP | undefined { return this.anp; }
}

/**
 * UAICP - Universal AI Communication Protocol
 */
export class UAICP extends EventEmitter {
    private agentId: string;
    private handlers: Map<string, MessageHandler>;
    
    // Cryptographic keys
    private signKeyPair: crypto.KeyPairKeyObjectResult;
    
    constructor(agentId: string) {
        super();
        this.agentId = agentId;
        this.handlers = new Map();
        
        // Generate Ed25519 key pair
        this.signKeyPair = crypto.generateKeyPairSync('ed25519');
    }
    
    async start(): Promise<void> {
        // Initialize protocol
    }
    
    async stop(): Promise<void> {
        // Cleanup
    }
    
    async sendMessage(message: Message): Promise<Message> {
        // Sign message
        const messageData = Buffer.from(JSON.stringify(message));
        message.signature = crypto.sign(null, messageData, this.signKeyPair.privateKey);
        
        // Send message (implement actual sending)
        this.emit('messageSent', message);
        
        return message;
    }
    
    verifyMessage(message: Message): boolean {
        if (!message.signature) return false;
        
        const messageData = Buffer.from(JSON.stringify({
            ...message,
            signature: undefined
        }));
        
        // Verify signature (would need sender's public key)
        return true; // Simplified
    }
    
    registerHandler(messageType: string, handler: MessageHandler): void {
        this.handlers.set(messageType, handler);
    }
}

/**
 * ANP - Agent Network Protocol
 */
export class ANP {
    private agentId: string;
    private services: Map<string, ServiceRegistration>;
    
    constructor(agentId: string) {
        this.agentId = agentId;
        this.services = new Map();
    }
    
    async start(): Promise<void> {
        // Initialize service discovery
    }
    
    async stop(): Promise<void> {
        // Cleanup
    }
    
    registerService(service: ServiceRegistration): string {
        const serviceId = uuidv4();
        this.services.set(serviceId, service);
        return serviceId;
    }
    
    discoverServices(capabilities: string[]): ServiceRegistration[] {
        return Array.from(this.services.values()).filter(service =>
            capabilities.every(cap => service.capabilities.includes(cap))
        );
    }
}

/**
 * Legacy Bridge for mainframe integration
 */
export class LegacyBridge {
    private config: VeloraConfig;
    private iso20022Handler?: ISO20022Handler;
    private cobolParser?: COBOLParser;
    private cicsGateway?: CICSGateway;
    
    constructor(config: VeloraConfig) {
        this.config = config;
    }
    
    async initialize(): Promise<void> {
        this.iso20022Handler = new ISO20022Handler();
        this.cobolParser = new COBOLParser();
        this.cicsGateway = new CICSGateway(
            this.config.mainframeHost,
            this.config.mainframePort
        );
    }
    
    async start(): Promise<void> {
        await this.cicsGateway?.connect();
    }
    
    async stop(): Promise<void> {
        await this.cicsGateway?.disconnect();
    }
    
    /**
     * Process ISO 20022 message to legacy system
     */
    async processISO20022ToLegacy(isoMessage: string): Promise<TransactionResult> {
        // Parse ISO 20022
        const message = await this.iso20022Handler!.parse(isoMessage);
        
        // Convert to COBOL
        const cobolData = this.cobolParser!.convertToCobol(message);
        
        // Execute CICS transaction
        return await this.cicsGateway!.executeTransaction('PAYM', cobolData);
    }
    
    /**
     * Process REST request to CICS
     */
    async processRestToCICS(restData: any, transactionId: string): Promise<any> {
        // Convert to COBOL format
        const cobolData = this.cobolParser!.jsonToCobol(restData);
        
        // Execute CICS transaction
        const result = await this.cicsGateway!.executeTransaction(transactionId, cobolData);
        
        // Convert response back to JSON
        return this.cobolParser!.cobolToJson(result.responseData);
    }
}

/**
 * ISO 20022 Message Handler
 */
export class ISO20022Handler {
    private parser: xml2js.Parser;
    private builder: xml2js.Builder;
    
    constructor() {
        this.parser = new xml2js.Parser();
        this.builder = new xml2js.Builder();
    }
    
    async parse(xmlContent: string): Promise<ISO20022Message> {
        const result = await this.parser.parseStringPromise(xmlContent);
        
        // Extract fields from parsed XML
        const doc = result.Document;
        const grpHdr = doc.CstmrCdtTrfInitn[0].GrpHdr[0];
        
        return {
            messageId: grpHdr.MsgId[0],
            creationDateTime: new Date(grpHdr.CreDtTm[0]),
            numberOfTransactions: parseInt(grpHdr.NbOfTxs[0]),
            controlSum: parseFloat(grpHdr.CtrlSum[0]),
            initiatingParty: grpHdr.InitgPty[0].Nm[0],
            paymentInstructions: this.extractPaymentInstructions(doc)
        };
    }
    
    generate(message: ISO20022Message): string {
        const xml = {
            Document: {
                $: { xmlns: 'urn:iso:std:iso:20022:tech:xsd:pain.001.001.03' },
                CstmrCdtTrfInitn: {
                    GrpHdr: {
                        MsgId: message.messageId,
                        CreDtTm: message.creationDateTime.toISOString(),
                        NbOfTxs: message.numberOfTransactions,
                        CtrlSum: message.controlSum,
                        InitgPty: { Nm: message.initiatingParty }
                    },
                    PmtInf: message.paymentInstructions
                }
            }
        };
        
        return this.builder.buildObject(xml);
    }
    
    private extractPaymentInstructions(doc: any): any[] {
        // Extract payment instructions from document
        return doc.CstmrCdtTrfInitn[0].PmtInf || [];
    }
}

/**
 * COBOL Copybook Parser
 */
export class COBOLParser {
    private ebcdicEncoding = 'cp037';
    
    /**
     * Convert ISO 20022 message to COBOL format
     */
    convertToCobol(message: ISO20022Message): Buffer {
        const record = Buffer.alloc(500); // Fixed record length
        
        // Write fields according to copybook layout
        this.writeField(record, 0, 20, message.messageId);
        this.writeField(record, 20, 8, this.formatDate(message.creationDateTime));
        this.writeField(record, 28, 6, this.formatTime(message.creationDateTime));
        
        // Write amount as packed decimal
        this.writePackedDecimal(record, 50, 8, Math.floor(message.controlSum * 100));
        
        return record;
    }
    
    /**
     * Convert JSON to COBOL format
     */
    jsonToCobol(data: any): Buffer {
        const record = Buffer.alloc(500);
        
        // Map JSON fields to COBOL layout
        if (data.transactionId) {
            this.writeField(record, 0, 20, data.transactionId);
        }
        if (data.amount) {
            this.writePackedDecimal(record, 50, 8, Math.floor(data.amount * 100));
        }
        if (data.accountNumber) {
            this.writeField(record, 100, 34, data.accountNumber);
        }
        
        return record;
    }
    
    /**
     * Convert COBOL data to JSON
     */
    cobolToJson(data: Buffer): any {
        return {
            transactionId: this.readField(data, 0, 20),
            status: this.readField(data, 20, 2),
            amount: this.readPackedDecimal(data, 50, 8) / 100,
            accountNumber: this.readField(data, 100, 34)
        };
    }
    
    private writeField(buffer: Buffer, offset: number, length: number, value: string): void {
        const encoded = iconv.encode(value.padEnd(length), this.ebcdicEncoding);
        encoded.copy(buffer, offset, 0, Math.min(encoded.length, length));
    }
    
    private readField(buffer: Buffer, offset: number, length: number): string {
        const slice = buffer.slice(offset, offset + length);
        return iconv.decode(slice, this.ebcdicEncoding).trim();
    }
    
    private writePackedDecimal(buffer: Buffer, offset: number, length: number, value: number): void {
        const digits = Math.abs(value).toString().padStart(length * 2 - 1, '0');
        const packed = Buffer.alloc(length);
        
        // Pack digits
        for (let i = 0; i < digits.length - 1; i += 2) {
            const high = parseInt(digits[i]);
            const low = parseInt(digits[i + 1]);
            packed[Math.floor(i / 2)] = (high << 4) | low;
        }
        
        // Add sign nibble
        const lastDigit = parseInt(digits[digits.length - 1]);
        packed[length - 1] = (lastDigit << 4) | (value >= 0 ? 0x0C : 0x0D);
        
        packed.copy(buffer, offset);
    }
    
    private readPackedDecimal(buffer: Buffer, offset: number, length: number): number {
        const packed = buffer.slice(offset, offset + length);
        let digits = '';
        
        // Unpack digits
        for (let i = 0; i < length - 1; i++) {
            digits += ((packed[i] >> 4) & 0x0F).toString();
            digits += (packed[i] & 0x0F).toString();
        }
        
        // Last byte has digit and sign
        digits += ((packed[length - 1] >> 4) & 0x0F).toString();
        const sign = (packed[length - 1] & 0x0F) === 0x0D ? -1 : 1;
        
        return parseInt(digits) * sign;
    }
    
    private formatDate(date: Date): string {
        return date.toISOString().slice(0, 10).replace(/-/g, '');
    }
    
    private formatTime(date: Date): string {
        return date.toISOString().slice(11, 19).replace(/:/g, '');
    }
}

/**
 * CICS Transaction Gateway
 */
export class CICSGateway {
    private host: string;
    private port: number;
    private client?: net.Socket;
    
    constructor(host: string, port: number) {
        this.host = host;
        this.port = port;
    }
    
    async connect(): Promise<void> {
        return new Promise((resolve, reject) => {
            this.client = new net.Socket();
            
            this.client.connect(this.port, this.host, () => {
                console.log(`Connected to CICS at ${this.host}:${this.port}`);
                resolve();
            });
            
            this.client.on('error', reject);
        });
    }
    
    async disconnect(): Promise<void> {
        return new Promise((resolve) => {
            if (this.client) {
                this.client.end(() => {
                    this.client = undefined;
                    resolve();
                });
            } else {
                resolve();
            }
        });
    }
    
    async executeTransaction(transactionId: string, commarea: Buffer): Promise<TransactionResult> {
        return new Promise((resolve, reject) => {
            if (!this.client) {
                reject(new Error('Not connected to CICS'));
                return;
            }
            
            // Build ECI request
            const request = this.buildECIRequest(transactionId, commarea);
            
            // Send request
            this.client.write(request);
            
            // Wait for response
            this.client.once('data', (data) => {
                const result = this.parseECIResponse(data);
                resolve(result);
            });
            
            this.client.once('error', reject);
        });
    }
    
    private buildECIRequest(transactionId: string, commarea: Buffer): Buffer {
        const header = Buffer.alloc(32);
        
        // ECI header
        header.write('ECI ', 0, 4);
        header.writeUInt16BE(1, 4); // Version
        header.writeUInt16BE(commarea.length, 6);
        header.write(transactionId.padEnd(8), 8, 8);
        
        return Buffer.concat([header, commarea]);
    }
    
    private parseECIResponse(data: Buffer): TransactionResult {
        const responseCode = data.readUInt8(6);
        const responseData = data.slice(32);
        
        return {
            success: responseCode === 0,
            responseCode: responseCode.toString().padStart(2, '0'),
            responseData
        };
    }
}

// Type definitions

export interface VeloraConfig {
    agentId: string;
    environment: string;
    version: string;
    mainframeHost: string;
    mainframePort: number;
    
    static getDefault(): VeloraConfig;
}

export interface Message {
    messageId: string;
    messageType: string;
    sourceAgent: string;
    destAgent: string;
    payload: any;
    timestamp: Date;
    signature?: Buffer;
}

export interface ISO20022Message {
    messageId: string;
    creationDateTime: Date;
    numberOfTransactions: number;
    controlSum: number;
    initiatingParty: string;
    paymentInstructions: any[];
}

export interface TransactionResult {
    success: boolean;
    responseCode: string;
    responseData: Buffer;
}

export interface ServiceRegistration {
    serviceId: string;
    serviceName: string;
    capabilities: string[];
    endpoint?: string;
}

export interface HealthCheckResult {
    status: string;
    instanceId: string;
    uptime: number;
    components: Record<string, string>;
}

export interface SystemStatus {
    instanceId: string;
    isRunning: boolean;
    startTime: Date;
    uptime: number;
    environment: string;
    version: string;
}

export type MessageHandler = (message: Message) => Promise<void>;

export class VeloraException extends Error {
    constructor(message: string, public cause?: any) {
        super(message);
        this.name = 'VeloraException';
    }
}

// Stub implementations for other managers
export class AgentManager {
    constructor(private config: VeloraConfig, private pm: ProtocolManager) {}
    async initialize(): Promise<void> {}
    async start(): Promise<void> {}
    async stop(): Promise<void> {}
}

export class SecurityManager {
    constructor(private config: VeloraConfig) {}
    async initialize(): Promise<void> {}
    async start(): Promise<void> {}
    async stop(): Promise<void> {}
}

export class DataManager {
    constructor(private config: VeloraConfig) {}
    async initialize(): Promise<void> {}
    async start(): Promise<void> {}
    async stop(): Promise<void> {}
}

// Export main class as default
export default VeloraCore;

File: velora/core/velora.py
"""
Main Velora Core class that orchestrates the entire system
"""

import asyncio
from typing import Optional, Dict, Any, List
from datetime import datetime
import uuid

from velora.core.config import Settings, get_settings
from velora.core.logging import LoggerMixin, setup_logging
from velora.core.exceptions import VeloraException, ConfigurationError


class VeloraCore(LoggerMixin):
    """
    Main Velora system orchestrator
    
    This class manages the lifecycle of all Velora components including:
    - Protocol handlers (UAICP, ANP)
    - Agent management
    - Data layer
    - Security framework
    - API gateway
    - Legacy integrations
    """
    
    def __init__(self, settings: Optional[Settings] = None):
        """
        Initialize Velora Core
        
        Args:
            settings: Optional settings override
        """
        self.settings = settings or get_settings()
        self.instance_id = str(uuid.uuid4())
        self.start_time = datetime.utcnow()
        self.is_running = False
        
        # Component references (will be initialized in setup)
        self.protocol_manager = None
        self.agent_manager = None
        self.data_manager = None
        self.security_manager = None
        self.api_gateway = None
        self.legacy_bridge = None
        self.semantic_engine = None
        self.monitoring = None
        
        # Setup logging
        setup_logging(self.settings.log_level)
        
        self.log_info(
            "Velora Core initialized",
            instance_id=self.instance_id,
            environment=self.settings.environment
        )
    
    async def setup(self) -> None:
        """
        Set up all Velora components
        """
        try:
            self.log_info("Starting Velora setup")
            
            # Initialize protocol manager
            await self._setup_protocols()
            
            # Initialize agent manager
            await self._setup_agents()
            
            # Initialize data layer
            await self._setup_data_layer()
            
            # Initialize security framework
            await self._setup_security()
            
            # Initialize API gateway
            await self._setup_api_gateway()
            
            # Initialize legacy bridge
            await self._setup_legacy_bridge()
            
            # Initialize semantic engine
            await self._setup_semantic_engine()
            
            # Initialize monitoring
            await self._setup_monitoring()
            
            self.log_info("Velora setup completed successfully")
            
        except Exception as e:
            self.log_error("Failed to setup Velora", error=e)
            raise ConfigurationError(f"Setup failed: {str(e)}")
    
    async def _setup_protocols(self) -> None:
        """Initialize protocol handlers"""
        from velora.protocols import ProtocolManager
        
        self.log_info("Setting up protocol manager")
        self.protocol_manager = ProtocolManager(self.settings)
        await self.protocol_manager.initialize()
    
    async def _setup_agents(self) -> None:
        """Initialize agent management system"""
        from velora.agents import AgentManager
        
        self.log_info("Setting up agent manager")
        self.agent_manager = AgentManager(self.settings, self.protocol_manager)
        await self.agent_manager.initialize()
    
    async def _setup_data_layer(self) -> None:
        """Initialize data layer"""
        from velora.data import DataManager
        
        self.log_info("Setting up data layer")
        self.data_manager = DataManager(self.settings)
        await self.data_manager.initialize()
    
    async def _setup_security(self) -> None:
        """Initialize security framework"""
        from velora.security import SecurityManager
        
        self.log_info("Setting up security framework")
        self.security_manager = SecurityManager(self.settings)
        await self.security_manager.initialize()
    
    async def _setup_api_gateway(self) -> None:
        """Initialize API gateway"""
        from velora.api import APIGateway
        
        self.log_info("Setting up API gateway")
        self.api_gateway = APIGateway(
            self.settings,
            self.protocol_manager,
            self.agent_manager,
            self.security_manager
        )
        await self.api_gateway.initialize()
    
    async def _setup_legacy_bridge(self) -> None:
        """Initialize legacy system bridge"""
        from velora.legacy import LegacyBridge
        
        self.log_info("Setting up legacy bridge")
        self.legacy_bridge = LegacyBridge(self.settings, self.data_manager)
        await self.legacy_bridge.initialize()
    
    async def _setup_semantic_engine(self) -> None:
        """Initialize semantic knowledge engine"""
        from velora.semantics import SemanticEngine
        
        self.log_info("Setting up semantic engine")
        self.semantic_engine = SemanticEngine(self.settings, self.data_manager)
        await self.semantic_engine.initialize()
    
    async def _setup_monitoring(self) -> None:
        """Initialize monitoring and observability"""
        from velora.monitoring import MonitoringSystem
        
        self.log_info("Setting up monitoring system")
        self.monitoring = MonitoringSystem(self.settings)
        await self.monitoring.initialize()
    
    async def start(self) -> None:
        """
        Start the Velora system
        """
        if self.is_running:
            self.log_warning("Velora is already running")
            return
        
        try:
            self.log_info("Starting Velora system")
            
            # Setup components if not already done
            if not self.protocol_manager:
                await self.setup()
            
            # Start all components
            await self.protocol_manager.start()
            await self.agent_manager.start()
            await self.data_manager.start()
            await self.security_manager.start()
            await self.api_gateway.start()
            await self.legacy_bridge.start()
            await self.semantic_engine.start()
            await self.monitoring.start()
            
            self.is_running = True
            self.log_info(
                "Velora system started successfully",
                instance_id=self.instance_id,
                uptime_seconds=self.get_uptime()
            )
            
        except Exception as e:
            self.log_error("Failed to start Velora", error=e)
            await self.stop()
            raise VeloraException(f"Start failed: {str(e)}")
    
    async def stop(self) -> None:
        """
        Stop the Velora system
        """
        if not self.is_running:
            self.log_warning("Velora is not running")
            return
        
        try:
            self.log_info("Stopping Velora system")
            
            # Stop all components in reverse order
            if self.monitoring:
                await self.monitoring.stop()
            if self.semantic_engine:
                await self.semantic_engine.stop()
            if self.legacy_bridge:
                await self.legacy_bridge.stop()
            if self.api_gateway:
                await self.api_gateway.stop()
            if self.security_manager:
                await self.security_manager.stop()
            if self.data_manager:
                await self.data_manager.stop()
            if self.agent_manager:
                await self.agent_manager.stop()
            if self.protocol_manager:
                await self.protocol_manager.stop()
            
            self.is_running = False
            self.log_info(
                "Velora system stopped",
                instance_id=self.instance_id,
                total_uptime_seconds=self.get_uptime()
            )
            
        except Exception as e:
            self.log_error("Error during shutdown", error=e)
            raise VeloraException(f"Shutdown error: {str(e)}")
    
    async def restart(self) -> None:
        """
        Restart the Velora system
        """
        self.log_info("Restarting Velora system")
        await self.stop()
        await asyncio.sleep(2)  # Brief pause before restart
        await self.start()
    
    def get_status(self) -> Dict[str, Any]:
        """
        Get current system status
        
        Returns:
            Dictionary containing system status information
        """
        return {
            "instance_id": self.instance_id,
            "is_running": self.is_running,
            "start_time": self.start_time.isoformat(),
            "uptime_seconds": self.get_uptime(),
            "environment": self.settings.environment,
            "version": self.settings.app_version,
            "components": {
                "protocol_manager": self._get_component_status(self.protocol_manager),
                "agent_manager": self._get_component_status(self.agent_manager),
                "data_manager": self._get_component_status(self.data_manager),
                "security_manager": self._get_component_status(self.security_manager),
                "api_gateway": self._get_component_status(self.api_gateway),
                "legacy_bridge": self._get_component_status(self.legacy_bridge),
                "semantic_engine": self._get_component_status(self.semantic_engine),
                "monitoring": self._get_component_status(self.monitoring),
            }
        }
    
    def _get_component_status(self, component) -> str:
        """Get status of a component"""
        if component is None:
            return "not_initialized"
        elif hasattr(component, "is_running"):
            return "running" if component.is_running else "stopped"
        else:
            return "initialized"
    
    def get_uptime(self) -> float:
        """Get system uptime in seconds"""
        return (datetime.utcnow() - self.start_time).total_seconds()
    
    async def health_check(self) -> Dict[str, Any]:
        """
        Perform health check on all components
        
        Returns:
            Health check results
        """
        health = {
            "status": "healthy",
            "timestamp": datetime.utcnow().isoformat(),
            "instance_id": self.instance_id,
            "uptime_seconds": self.get_uptime(),
            "checks": {}
        }
        
        # Check each component
        components = [
            ("protocol_manager", self.protocol_manager),
            ("agent_manager", self.agent_manager),
            ("data_manager", self.data_manager),
            ("security_manager", self.security_manager),
            ("api_gateway", self.api_gateway),
            ("legacy_bridge", self.legacy_bridge),
            ("semantic_engine", self.semantic_engine),
            ("monitoring", self.monitoring),
        ]
        
        for name, component in components:
            if component and hasattr(component, "health_check"):
                try:
                    component_health = await component.health_check()
                    health["checks"][name] = component_health
                    if component_health.get("status") != "healthy":
                        health["status"] = "degraded"
                except Exception as e:
                    health["checks"][name] = {
                        "status": "unhealthy",
                        "error": str(e)
                    }
                    health["status"] = "unhealthy"
            else:
                health["checks"][name] = {
                    "status": "unknown" if component else "not_initialized"
                }
        
        return health